aips/aip-20.md

```diff
@@ -0,0 +1,467 @@
+
+    /// Hash an arbitrary-length byte array `msg` into structure `S` with a domain separation tag `dst`
+    /// using the given hash-to-structure suite `H`.
+    public fun hash_to<St, Su>(dst: &vector<u8>, msg: &vector<u8>): Element<St>;
+
+    #[test_only]
+    /// Generate a random element of an algebraic structure `S`.
+    public fun rand_insecure<S>(): Element<S>;
+
+```
+
+In general, every structure implements basic operations like (de)serialization, equality check, random sampling.
+
+For example, A group may also implement the following operations. (Additive notions are used.)
+- `order()` for getting the group order.
+- `zero()` for getting the group identity.
+- `one()` for getting the group generator (if exists).
+- `neg()` for group element inversion.
+- `add()` for basic group operation.
+- `sub()` for group element subtraction.
+- `double()` for efficient group element doubling.
+- `scalar_mul()` for group scalar multiplication.
+- `multi_scalar_mul()` for efficient group multi-scalar multiplication.
+- `hash_to()` for hash-to-group.
+
+As another example, a field may also implement the following operations.
+- `order()` for getting the field order.
+- `zero()` for the field additive identity.
+- `one()` for the field multiplicative identity.
+- `add()` for field addition.
+- `sub()` for field subtraction.
+- `mul()` for field multiplication.
+- `div()` for field division.
+- `neg()` for field negation.
+- `inv()` for field inversion.
+- `sqr()` for efficient field element squaring.
+- `from_u64()` for quick conversion from u64 to field element.
+
+Similarly, for 3 groups `G1`, `G2`, `Gt` that admit a bilinear map, `pairing<G1, G2, Gt>()` and `multi_pairing<G1, G2, Gt>()` may be implemented.
+
+For a subset/superset relationship between 2 structures, `upcast()` and `downcast()` may be implemented.
+E.g., in BLS12-381 `Gt` is a multiplicative subgroup from `Fq12` so upcasting from `Gt` to `Fq12` and downcasting from `Fq12` to `Gt` can be supported.
+
+#### Shared Scalar Fields
+
+Some groups share the same group order,
+and an ergonomic design for this is to
+allow multiple groups to share the same scalar field
+(mainly for the purpose of scalar multiplication),
+if they have the same order.
+
+In other words, the following should be supported.
+
+```rust
+// algebra.move
+pub fun scalar_mul<G,S>(element: &Element<G>, scalar: &Scalar<S>)...
+
+// user_contract.move
+**let k: Scalar<ScalarForBx> = somehow_get_k();**
+let p1 = one<GroupB1>();
+let p2 = one<GroupB2>();
+let q1 = scalar_mul<GroupB1, ScalarForBx>(&p1, &k);
+let q2 = scalar_mul<GroupB2, ScalarForBx>(&p2, &k);
+```
+
+#### Handling Incorrect Type Parameter(s)
+
+There is currently no easy way to ensure type safety for the generic operations.
+E.g., `pairing<A,B,C>(a,b,c)` can compile even there is no pairing between `A,B,C` existing/implemented.
+
+Therefore the backend should handle the type checks at runtime.
```

Comment:
Typo: Therefore**,** the backend **must** handle.

Reply:
Fixed.